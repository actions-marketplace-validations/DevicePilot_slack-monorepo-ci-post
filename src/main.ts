import * as core from '@actions/core'
import {PostMessageInputs, postMessage} from './post-message'

function asStatus(str: string): PostMessageInputs['status'] {
  const statuses: PostMessageInputs['status'][] = [
    'neutral',
    'success',
    'cancelled',
    'timed_out',
    'failure'
  ]
  if (!statuses.includes(str as PostMessageInputs['status'])) {
    throw new TypeError('status must be success or failure')
  }
  return str as PostMessageInputs['status']
}

function viaEnv(str: string): string {
  if (!(str.startsWith('${') && str.endsWith('}'))) {
    return str
  }
  const name = str.substring(2, str.length - 1)
  const value = process.env[name]
  return value || str
}

async function run(): Promise<void> {
  try {
    const inputs: PostMessageInputs = {
      status: asStatus(core.getInput('status')),
      scopes: JSON.parse(core.getInput('scopes')).scope,
      version: core.getInput('version'),
      userIds: JSON.parse(viaEnv(core.getInput('userIds'))),
      slackToken: viaEnv(core.getInput('slackToken')),
      conversationId: viaEnv(core.getInput('conversationId')),
      repository: core.getInput('repository'),
      stage: core.getInput('stage'),
      commits: JSON.parse(core.getInput('commits')),
      runUrl: core.getInput('runUrl'),
      sourceUrl: core.getInput('sourceUrl'),
      username: core.getInput('username')
    }
    await postMessage(inputs)
  } catch (error) {
    core.setFailed(error.message)
  }
}

run()
