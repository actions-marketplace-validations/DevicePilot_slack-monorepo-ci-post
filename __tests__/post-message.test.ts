import {postMessage, PostMessageInputs} from '../src/post-message'

const mockPostMessage = jest.fn().mockResolvedValue({})
jest.mock('@slack/web-api', () => ({
  WebClient: jest.fn().mockImplementation(() => {
    return {chat: {postMessage: mockPostMessage}}
  })
}))

test('posts a simple message', async () => {
  const input: PostMessageInputs = {
    commits: ['fix: works perfectly\nfixes org/project#123'],
    conversationId: 'conversation_id',
    repository: 'org/project',
    runUrl: 'example.com/actions/this_run',
    scopes: [...Array(12)].map((_, i) => `@org/project-${i}`),
    slackToken: 'my_slack_token',
    sourceUrl: 'example.com/sources/this_push',
    stage: 'development',
    status: 'success',
    userIds: {user_one: 'slack_user_one', user_two: 'slack_user_two'},
    username: 'user_two',
    version: undefined
  }
  await postMessage(input)
  expect(mockPostMessage).toMatchSnapshot()
})
