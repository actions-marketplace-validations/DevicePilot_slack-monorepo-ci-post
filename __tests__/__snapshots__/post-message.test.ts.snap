// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts a simple message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "*org/project* _development_ has completed :white_check_mark:",
              "type": "mrkdwn",
            },
            "type": "section",
          },
          Object {
            "fields": Array [
              Object {
                "text": "_@org_/*project-0*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-1*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-2*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-3*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-4*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-5*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-6*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-7*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-8*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-9*",
                "type": "mrkdwn",
              },
            ],
            "type": "section",
          },
          Object {
            "fields": Array [
              Object {
                "text": "_@org_/*project-10*",
                "type": "mrkdwn",
              },
              Object {
                "text": "_@org_/*project-11*",
                "type": "mrkdwn",
              },
            ],
            "type": "section",
          },
          Object {
            "text": Object {
              "text": "*fix*: works perfectly, fixes <https://github.com/org/project/issues/123|#123>",
              "type": "mrkdwn",
            },
            "type": "section",
          },
          Object {
            "elements": Array [
              Object {
                "text": "<@slack_user_two>",
                "type": "mrkdwn",
              },
            ],
            "type": "context",
          },
          Object {
            "elements": Array [
              Object {
                "style": "primary",
                "text": Object {
                  "text": "Build Log",
                  "type": "plain_text",
                },
                "type": "button",
                "url": "example.com/actions/this_run",
              },
              Object {
                "text": Object {
                  "text": "Source",
                  "type": "plain_text",
                },
                "type": "button",
                "url": "example.com/sources/this_push",
              },
            ],
            "type": "actions",
          },
        ],
        "channel": "conversation_id",
        "text": "org/project development: success",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
